// astro/dist/core/dev/restart.js
import { fileURLToPath } from "node:url";
import * as vite from "vite";
import { eventCliSession, telemetry } from "../../events/index.js";
import { createNodeLogging, createSettings, resolveConfig } from "../config/index.js";
import { collectErrorMetadata } from "../errors/dev/utils.js";
import { isAstroConfigZodError } from "../errors/errors.js";
import { createSafeError } from "../errors/index.js";
import { info, error as logError } from "../logger/core.js";
import { formatErrorMessage } from "../messages.js";
import { createContainer, isStarted, startContainer } from "./container.js";
async function createRestartedContainer(container, settings, needsStart) {
  const { logging, fs, inlineConfig } = container;
  const newContainer = await createContainer({
    isRestart: true,
    logging,
    settings,
    inlineConfig,
    fs
  });
  if (needsStart) {
    await startContainer(newContainer);
  }
  return newContainer;
}
function shouldRestartContainer({ settings, inlineConfig, restartInFlight }, changedFile) {
  if (restartInFlight)
    return false;
  let shouldRestart = false;
  if (inlineConfig.configFile) {
    shouldRestart = vite.normalizePath(inlineConfig.configFile) === vite.normalizePath(changedFile);
  } else {
    const exp = new RegExp(`.*astro.config.((mjs)|(cjs)|(js)|(ts))$`);
    const normalizedChangedFile = vite.normalizePath(changedFile);
    shouldRestart = exp.test(normalizedChangedFile);
  }
  if (!shouldRestart && settings.watchFiles.length > 0) {
    shouldRestart = settings.watchFiles.some(
      (path) => vite.normalizePath(path) === vite.normalizePath(changedFile)
    );
  }
  return shouldRestart;
}
async function restartContainer(container) {
  const { logging, close, settings: existingSettings } = container;
  container.restartInFlight = true;
  const needsStart = isStarted(container);
  try {
    const { astroConfig } = await resolveConfig(container.inlineConfig, "dev", container.fs);
    const settings = createSettings(astroConfig, fileURLToPath(existingSettings.config.root));
    await close();
    return {
      container: await createRestartedContainer(container, settings, needsStart),
      error: null
    };
  } catch (_err) {
    const error = createSafeError(_err);
    if (!isAstroConfigZodError(_err)) {
      logError(logging, "config", formatErrorMessage(collectErrorMetadata(error)) + "\n");
    }
    container.viteServer.ws.send({
      type: "error",
      err: {
        message: error.message,
        stack: error.stack || ""
      }
    });
    await close();
    info(logging, "astro", "Continuing with previous valid configuration\n");
    return {
      container: await createRestartedContainer(container, existingSettings, needsStart),
      error
    };
  }
}
async function createContainerWithAutomaticRestart({
  inlineConfig,
  fs
}) {
  const logging = createNodeLogging(inlineConfig ?? {});
  const { userConfig, astroConfig } = await resolveConfig(inlineConfig ?? {}, "dev", fs);
  telemetry.record(eventCliSession("dev", userConfig));
  const settings = createSettings(astroConfig, fileURLToPath(astroConfig.root));
  const initialContainer = await createContainer({ settings, logging, inlineConfig, fs });
  let resolveRestart;
  let restartComplete = new Promise((resolve) => {
    resolveRestart = resolve;
  });
  let restart = {
    container: initialContainer,
    restarted() {
      return restartComplete;
    }
  };
  async function handleServerRestart(logMsg) {
    info(logging, "astro", logMsg + "\n");
    const container = restart.container;
    const { container: newContainer, error } = await restartContainer(container);
    restart.container = newContainer;
    addWatches();
    resolveRestart(error);
    restartComplete = new Promise((resolve) => {
      resolveRestart = resolve;
    });
  }
  function handleChangeRestart(logMsg) {
    return async function(changedFile) {
      if (shouldRestartContainer(restart.container, changedFile)) {
        handleServerRestart(logMsg);
      }
    };
  }
  function addWatches() {
    const watcher = restart.container.viteServer.watcher;
    watcher.on("change", handleChangeRestart("Configuration updated. Restarting..."));
    watcher.on("unlink", handleChangeRestart("Configuration removed. Restarting..."));
    watcher.on("add", handleChangeRestart("Configuration added. Restarting..."));
    restart.container.viteServer.restart = () => handleServerRestart("Restarting...");
  }
  addWatches();
  return restart;
}
export {
  createContainerWithAutomaticRestart,
  restartContainer,
  shouldRestartContainer
};
